torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1979,6 +1983,906 @@ def clip_grad_norm_(
+        return flat_param_id_to_param  # type: ignore[return-value]
+
+    @staticmethod
+    def _preprocess_optim_state_dict(
+        optim_state_dict: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """
+        Pre-processes the rank's optimizer state dict by converting any
+        "step" state from a zero-dimension tensor to a float to support
+        backward compatibility with optimizer state dicts saved before version
+        1.12.
+
+        Args:
+            optim_state_dict (Dict[str, Any]): Optimizer state dict returned
+                from ``optim.state_dict()``, where ``optim`` is a rank's
+                optimizer.
+
+        Returns:
+            optim_state_dict (Dict[str, Any]): Pre-processed optimizer state
+                dict.
+
+        Raises:
+            ValueError: If there are any zero-dimension tensor states other
+                than "step" since we do not currently support them.
+        """
+        # Convert Adam's "step" state from a zero-dimension tensor to float if
+        # needed to support BC with dicts from before version 1.12
+        for param_state in optim_state_dict["state"].values():
+            if "step" in param_state:
+                step = param_state["step"]
+                if _is_zero_dim_tensor(step):
+                    param_state["step"] = step.item()
+        # Currently, we do not know of any other zero-dimension tensor state,
+        # so here, we check for such cases and error since it is unsupported;
+        # we can remove this if we choose to support zero-dimension tensors
+        for param_state in optim_state_dict["state"].values():
+            for state_name, state_value in param_state.items():
+                if state_name == "step":
+                    continue
+                if _is_zero_dim_tensor(state_value):
+                    raise ValueError(
+                        "Unsupported zero-dimension tensor state: "
+                        f"{state_name}"
+                    )
+        keys = set(optim_state_dict.keys())
+        assert keys == {"state", "param_groups"}, \
+            "Expects \"state\" and \"param_groups\" to be the keys in the " \
+            f"optimizer state dict but got {keys}"
+        return optim_state_dict
+
+    @staticmethod
+    def _unflatten_optim_state(
+        flat_param: FlatParameter,
+        flat_param_state: Dict[str, Any],
+        fsdp_module: "FullyShardedDataParallel",
+        target_rank: int,
+    ) -> Dict[int, Dict[str, Any]]:
+        """
+        Args:
+            flat_param (FlatParameter): Flattened parameter.
+            flat_param_state (Dict[str, Any]): Entry for a flattened parameter
+                in the "state" part of the optimizer state dict.
+            fsdp_module (FullyShardedDataParallel): FSDP module that owns the
+                flattened parameter, i.e. holds it in ``self.params``.
+
+        Returns:
+            unflat_param_state (Dict[int, Dict[str, Any]]): A :class:`dict`
```

Comment:
Yup. Good catch.

Reply:
We can simplify this to a `List[Dict[str, Any]]`. I think we still need one `Dict[str, Any]` per unflattened parameter since we should keep unflattened parameter's state separate.